File: macOS/TokenMihariban/OllamaProxyService.swift
import Foundation
import Network
import ClaudeUsageCore

enum OllamaProxyState: Equatable {
    case stopped
    case starting
    case running
    case failed(String)

    var isFailure: Bool {
        if case .failed = self { return true }
        return false
    }
}

/// Opt-in HTTP/1.1 proxy used because Ollama has per-response metrics but no cumulative
/// usage-history endpoint. It never persists request or response text; only the final
/// token counters parsed into `OllamaUsageEvent` leave the connection object.
final class OllamaProxyService: @unchecked Sendable {
    static let localPort: UInt16 = 11435
    static let cloudPort: UInt16 = 11436

    private let queue = DispatchQueue(label: "com.yoyoi441.TokenMihariban.ollama-proxy")
    private let eventHandler: @Sendable (OllamaUsageEvent) -> Void
    private let stateHandler: @Sendable (OllamaProxyState) -> Void
    private var listeners: [NWListener] = []
    private var readyCount = 0

    init(
        eventHandler: @escaping @Sendable (OllamaUsageEvent) -> Void,
        stateHandler: @escaping @Sendable (OllamaProxyState) -> Void
    ) {
        self.eventHandler = eventHandler
        self.stateHandler = stateHandler
    }

    func start() {
        queue.async { [weak self] in
            guard let self, self.listeners.isEmpty else { return }
            self.stateHandler(.starting)
            do {
                try self.startListener(
                    port: Self.localPort,
                    backend: URL(string: "http://127.0.0.1:11434")!,
                    forceCloud: false
                )
                try self.startListener(
                    port: Self.cloudPort,
                    backend: URL(string: "https://ollama.com")!,
                    forceCloud: true
                )
            } catch {
                self.stopLocked()
                self.stateHandler(.failed(error.localizedDescription))
            }
        }
    }

    func stop() {
        queue.async { [weak self] in
            self?.stopLocked()
            self?.stateHandler(.stopped)
        }
    }

    private func startListener(port: UInt16, backend: URL, forceCloud: Bool) throws {
        let parameters = NWParameters.tcp
        parameters.requiredLocalEndpoint = .hostPort(host: "127.0.0.1", port: NWEndpoint.Port(rawValue: port)!)
        let listener = try NWListener(using: parameters)
        listener.newConnectionHandler = { [eventHandler] connection in
            let handler = OllamaProxyConnection(
                connection: connection,
                backend: backend,
                forceCloud: forceCloud,
                eventHandler: eventHandler
            )
            handler.start()
        }
        listener.stateUpdateHandler = { [weak self] state in
            guard let self else { return }
            self.queue.async {
                switch state {
                case .ready:
                    self.readyCount += 1
                    if self.readyCount == 2 { self.stateHandler(.running) }
                case .failed(let error):
                    self.stateHandler(.failed(error.localizedDescription))
                default:
                    break
                }
            }
        }
        listeners.append(listener)
        listener.start(queue: queue)
    }

    private func stopLocked() {
        listeners.forEach { $0.cancel() }
        listeners.removeAll()
        readyCount = 0
    }

    deinit { listeners.forEach { $0.cancel() } }
}

private final class OllamaProxyConnection: NSObject, URLSessionDataDelegate, @unchecked Sendable {
    private struct ParsedRequest {
        let method: String
        let path: String
        let headers: [(String, String)]
        let body: Data
    }

    private let connection: NWConnection
    private let backend: URL
    private let forceCloud: Bool
    private let eventHandler: @Sendable (OllamaUsageEvent) -> Void
    private let requestId = UUID().uuidString
    private var incoming = Data()
    private var capturedResponse = Data()
    private var requestedModel: String?
    private var responseStarted = false
    private var session: URLSession?
    private var selfRetain: OllamaProxyConnection?

    init(
        connection: NWConnection,
        backend: URL,
        forceCloud: Bool,
        eventHandler: @escaping @Sendable (OllamaUsageEvent) -> Void
    ) {
        self.connection = connection
        self.backend = backend
        self.forceCloud = forceCloud
        self.eventHandler = eventHandler
    }

    func start() {
        selfRetain = self
        connection.stateUpdateHandler = { [weak self] state in
            switch state {
            case .ready: self?.receiveRequest()
            case .failed, .cancelled: self?.finish()
            default: break
            }
        }
        connection.start(queue: DispatchQueue.global(qos: .utility))
    }

    private func receiveRequest() {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 1_048_576) { [weak self] data, _, isComplete, error in
            guard let self else { return }
            if let data { self.incoming.append(data) }
            if self.incoming.count > 16 * 1_048_576 {
                self.sendError(status: 413, message: "Request too large")
                return
            }
            if let parsed = self.parseRequest() {
                self.forward(parsed)
            } else if error != nil || isComplete {
                self.sendError(status: 400, message: "Invalid HTTP request")
            } else {
                self.receiveRequest()
            }
        }
    }

    private func parseRequest() -> ParsedRequest? {
        let delimiter = Data([13, 10, 13, 10])
        guard let headerRange = incoming.range(of: delimiter),
              let headerText = String(data: incoming[..<headerRange.lowerBound], encoding: .utf8) else { return nil }
        let lines = headerText.components(separatedBy: "\r\n")
        guard let first = lines.first else { return nil }
        let requestParts = first.split(separator: " ", maxSplits: 2).map(String.init)
        guard requestParts.count == 3 else { return nil }

        var headers: [(String, String)] = []
        var contentLength = 0
        var chunked = false
        for line in lines.dropFirst() {
            guard let colon = line.firstIndex(of: ":") else { continue }
            let name = String(line[..<colon]).trimmingCharacters(in: .whitespaces)
            let value = String(line[line.index(after: colon)...]).trimmingCharacters(in: .whitespaces)
            headers.append((name, value))
            if name.caseInsensitiveCompare("Content-Length") == .orderedSame { contentLength = Int(value) ?? 0 }
            if name.caseInsensitiveCompare("Transfer-Encoding") == .orderedSame && value.lowercased().contains("chunked") { chunked = true }
        }
        if chunked {
            sendError(status: 411, message: "Chunked request bodies are not supported")
            return nil
        }
        let bodyStart = headerRange.upperBound
        guard incoming.count >= bodyStart + contentLength else { return nil }
        return ParsedRequest(
            method: requestParts[0],
            path: requestParts[1],
            headers: headers,
            body: incoming.subdata(in: bodyStart..<(bodyStart + contentLength))
        )
    }

    private func forward(_ parsed: ParsedRequest) {
        guard var components = URLComponents(url: backend, resolvingAgainstBaseURL: false) else {
            sendError(status: 502, message: "Invalid Ollama backend")
            return
        }
        let pathAndQuery = parsed.path.split(separator: "?", maxSplits: 1, omittingEmptySubsequences: false)
        components.path = String(pathAndQuery[0])
        components.percentEncodedQuery = pathAndQuery.count > 1 ? String(pathAndQuery[1]) : nil
        guard let url = components.url else {
            sendError(status: 400, message: "Invalid request path")
            return
        }

        if let object = try? JSONSerialization.jsonObject(with: parsed.body) as? [String: Any] {
            requestedModel = object["model"] as? String
        }

        var request = URLRequest(url: url)
        request.httpMethod = parsed.method
        request.httpBody = parsed.body.isEmpty ? nil : parsed.body
        for (name, value) in parsed.headers where !Self.hopByHopHeaders.contains(name.lowercased()) && name.caseInsensitiveCompare("Host") != .orderedSame {
            request.setValue(value, forHTTPHeaderField: name)
        }
        request.setValue("identity", forHTTPHeaderField: "Accept-Encoding")

        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 60 * 60
        let queue = OperationQueue()
        queue.maxConcurrentOperationCount = 1
        let session = URLSession(configuration: configuration, delegate: self, delegateQueue: queue)
        self.session = session
        session.dataTask(with: request).resume()
    }

    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse, completionHandler: @escaping (URLSession.ResponseDisposition) -> Void) {
        guard let http = response as? HTTPURLResponse else {
            completionHandler(.cancel)
            sendError(status: 502, message: "Invalid Ollama response")
            return
        }
        var text = "HTTP/1.1 \(http.statusCode) \(HTTPURLResponse.localizedString(forStatusCode: http.statusCode))\r\n"
        for (rawName, rawValue) in http.allHeaderFields {
            let name = String(describing: rawName)
            guard !Self.hopByHopHeaders.contains(name.lowercased()), name.caseInsensitiveCompare("Content-Length") != .orderedSame else { continue }
            text += "\(name): \(rawValue)\r\n"
        }
        text += "Transfer-Encoding: chunked\r\nConnection: close\r\n\r\n"
        responseStarted = true
        send(Data(text.utf8))
        completionHandler(.allow)
    }

    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
        if capturedResponse.count + data.count <= 128 * 1_048_576 { capturedResponse.append(data) }
        var chunk = Data(String(format: "%X\r\n", data.count).utf8)
        chunk.append(data)
        chunk.append(Data([13, 10]))
        send(chunk)
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        if let event = OllamaUsageResponseParser.parse(
            capturedResponse,
            requestedModel: requestedModel,
            forceCloud: forceCloud,
            requestId: requestId
        ) {
            eventHandler(event)
        }
        if responseStarted {
            send(Data("0\r\n\r\n".utf8), final: true)
        } else if let error {
            sendError(status: 502, message: error.localizedDescription)
        } else {
            sendError(status: 502, message: "Ollama did not return a response")
        }
    }

    private func send(_ data: Data, final: Bool = false) {
        connection.send(content: data, completion: .contentProcessed { [weak self] _ in
            if final { self?.finish() }
        })
    }

    private func sendError(status: Int, message: String) {
        let body = Data("{\"error\":\"\(message.replacingOccurrences(of: "\"", with: "'"))\"}".utf8)
        let response = "HTTP/1.1 \(status) Error\r\nContent-Type: application/json\r\nContent-Length: \(body.count)\r\nConnection: close\r\n\r\n"
        var data = Data(response.utf8)
        data.append(body)
        send(data, final: true)
    }

    private func finish() {
        session?.invalidateAndCancel()
        session = nil
        connection.cancel()
        selfRetain = nil
    }

    private static let hopByHopHeaders: Set<String> = [
        "connection", "keep-alive", "proxy-authenticate", "proxy-authorization",
        "te", "trailer", "transfer-encoding", "upgrade"
    ]
}

File: TokenMihariban/Ollama/OllamaProxyService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using TokenMihariban.Models;

namespace TokenMihariban.Ollama;

public enum OllamaProxyState { Stopped, Starting, Running, Failed }

/// <summary>
/// Loopback-only HTTP/1.1 proxy. Ollama exposes token metrics only in each response,
/// so an opt-in proxy is the only universal way to count CLI/editor/web-client usage
/// without storing prompt or response text.
/// </summary>
internal sealed class OllamaProxyService : IDisposable
{
    public const int LocalPort = 11435;
    public const int CloudPort = 11436;

    private readonly HttpClient _http = new(new SocketsHttpHandler { AutomaticDecompression = DecompressionMethods.None })
    {
        Timeout = Timeout.InfiniteTimeSpan
    };
    private CancellationTokenSource _stop = new();
    private readonly List<TcpListener> _listeners = new();
    private int _readyCount;

    public OllamaProxyState State { get; private set; } = OllamaProxyState.Stopped;
    public string? ErrorMessage { get; private set; }
    public event EventHandler<OllamaUsageEvent>? UsageCaptured;
    public event EventHandler? StateChanged;

    public void Start()
    {
        if (State is OllamaProxyState.Starting or OllamaProxyState.Running) return;
        if (_stop.IsCancellationRequested)
        {
            _stop.Dispose();
            _stop = new CancellationTokenSource();
        }
        State = OllamaProxyState.Starting;
        StateChanged?.Invoke(this, EventArgs.Empty);
        try
        {
            StartListener(LocalPort, new Uri("http://127.0.0.1:11434"), forceCloud: false);
            StartListener(CloudPort, new Uri("https://ollama.com"), forceCloud: true);
            State = OllamaProxyState.Running;
            StateChanged?.Invoke(this, EventArgs.Empty);
        }
        catch (Exception ex)
        {
            ErrorMessage = ex.Message;
            State = OllamaProxyState.Failed;
            foreach (var listener in _listeners) listener.Stop();
            _listeners.Clear();
            StateChanged?.Invoke(this, EventArgs.Empty);
        }
    }

    public void Stop()
    {
        if (State == OllamaProxyState.Stopped) return;
        _stop.Cancel();
        foreach (var listener in _listeners) listener.Stop();
        _listeners.Clear();
        _readyCount = 0;
        State = OllamaProxyState.Stopped;
        StateChanged?.Invoke(this, EventArgs.Empty);
    }

    private void StartListener(int port, Uri backend, bool forceCloud)
    {
        var listener = new TcpListener(IPAddress.Loopback, port);
        listener.Start();
        _listeners.Add(listener);
        _readyCount++;
        _ = AcceptLoopAsync(listener, backend, forceCloud, _stop.Token);
    }

    private async Task AcceptLoopAsync(TcpListener listener, Uri backend, bool forceCloud, CancellationToken cancellationToken)
    {
        while (!cancellationToken.IsCancellationRequested)
        {
            try
            {
                var client = await listener.AcceptTcpClientAsync(cancellationToken).ConfigureAwait(false);
                _ = HandleClientAsync(client, backend, forceCloud, cancellationToken);
            }
            catch (OperationCanceledException) { break; }
            catch (ObjectDisposedException) { break; }
            catch { if (!cancellationToken.IsCancellationRequested) await Task.Delay(250, cancellationToken).ConfigureAwait(false); }
        }
    }

    private async Task HandleClientAsync(TcpClient client, Uri backend, bool forceCloud, CancellationToken cancellationToken)
    {
        using (client)
        {
            var stream = client.GetStream();
            try
            {
                var request = await ReadRequestAsync(stream, cancellationToken).ConfigureAwait(false);
                if (request is null) return;
                if (request.IsChunked)
                {
                    await WriteErrorAsync(stream, 411, "Chunked request bodies are not supported", cancellationToken).ConfigureAwait(false);
                    return;
                }

                var target = new Uri(backend, request.Path);
                using var outgoing = new HttpRequestMessage(new HttpMethod(request.Method), target);
                if (request.Body.Length > 0) outgoing.Content = new ByteArrayContent(request.Body);
                foreach (var (name, value) in request.Headers)
                {
                    if (HopByHop.Contains(name) || name.Equals("Host", StringComparison.OrdinalIgnoreCase) || name.Equals("Content-Length", StringComparison.OrdinalIgnoreCase)) continue;
                    if (!outgoing.Headers.TryAddWithoutValidation(name, value) && outgoing.Content is not null)
                        outgoing.Content.Headers.TryAddWithoutValidation(name, value);
                }
                outgoing.Headers.Remove("Accept-Encoding");
                outgoing.Headers.TryAddWithoutValidation("Accept-Encoding", "identity");

                string? requestedModel = null;
                try
                {
                    using var json = JsonDocument.Parse(request.Body);
                    if (json.RootElement.TryGetProperty("model", out var model)) requestedModel = model.GetString();
                }
                catch { }

                using var response = await _http.SendAsync(outgoing, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
                await WriteResponseHeadersAsync(stream, response, cancellationToken).ConfigureAwait(false);
                await using var responseStream = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
                using var captured = new MemoryStream();
                var buffer = new byte[64 * 1024];
                while (true)
                {
                    var count = await responseStream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
                    if (count == 0) break;
                    if (captured.Length + count <= 128L * 1024 * 1024) captured.Write(buffer, 0, count);
                    await WriteChunkAsync(stream, buffer.AsMemory(0, count), cancellationToken).ConfigureAwait(false);
                }
                await stream.WriteAsync(Encoding.ASCII.GetBytes("0\r\n\r\n"), cancellationToken).ConfigureAwait(false);

                var usage = OllamaUsageResponseParser.Parse(captured.ToArray(), requestedModel, forceCloud);
                if (usage is not null) UsageCaptured?.Invoke(this, usage);
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                try { await WriteErrorAsync(stream, 502, ex.Message, cancellationToken).ConfigureAwait(false); }
                catch { }
            }
        }
    }

    private sealed record IncomingRequest(string Method, string Path, IReadOnlyList<(string Name, string Value)> Headers, byte[] Body, bool IsChunked);

    private static async Task<IncomingRequest?> ReadRequestAsync(NetworkStream stream, CancellationToken cancellationToken)
    {
        using var collected = new MemoryStream();
        var one = new byte[1];
        while (collected.Length < 64 * 1024)
        {
            var read = await stream.ReadAsync(one, cancellationToken).ConfigureAwait(false);
            if (read == 0) return null;
            collected.WriteByte(one[0]);
            if (collected.Length >= 4)
            {
                var bytes = collected.GetBuffer();
                var n = (int)collected.Length;
                if (bytes[n - 4] == 13 && bytes[n - 3] == 10 && bytes[n - 2] == 13 && bytes[n - 1] == 10) break;
            }
        }
        if (collected.Length >= 64 * 1024) throw new InvalidDataException("HTTP headers are too large");

        var headerText = Encoding.ASCII.GetString(collected.ToArray());
        var lines = headerText.Split("\r\n", StringSplitOptions.None);
        var requestLine = lines[0].Split(' ', 3);
        if (requestLine.Length != 3) throw new InvalidDataException("Invalid HTTP request line");
        var headers = new List<(string, string)>();
        var contentLength = 0;
        var chunked = false;
        foreach (var line in lines.Skip(1))
        {
            var colon = line.IndexOf(':');
            if (colon <= 0) continue;
            var name = line[..colon].Trim();
            var value = line[(colon + 1)..].Trim();
            headers.Add((name, value));
            if (name.Equals("Content-Length", StringComparison.OrdinalIgnoreCase)) int.TryParse(value, out contentLength);
            if (name.Equals("Transfer-Encoding", StringComparison.OrdinalIgnoreCase) && value.Contains("chunked", StringComparison.OrdinalIgnoreCase)) chunked = true;
        }
        if (contentLength > 16 * 1024 * 1024) throw new InvalidDataException("Request body is too large");
        var body = new byte[Math.Max(0, contentLength)];
        var offset = 0;
        while (offset < body.Length)
        {
            var read = await stream.ReadAsync(body.AsMemory(offset), cancellationToken).ConfigureAwait(false);
            if (read == 0) throw new EndOfStreamException();
            offset += read;
        }
        return new IncomingRequest(requestLine[0], requestLine[1], headers, body, chunked);
    }

    private static async Task WriteResponseHeadersAsync(NetworkStream stream, HttpResponseMessage response, CancellationToken cancellationToken)
    {
        var builder = new StringBuilder($"HTTP/1.1 {(int)response.StatusCode} {response.ReasonPhrase}\r\n");
        foreach (var header in response.Headers.Concat(response.Content.Headers))
        {
            if (HopByHop.Contains(header.Key) || header.Key.Equals("Content-Length", StringComparison.OrdinalIgnoreCase)) continue;
            foreach (var value in header.Value) builder.Append(header.Key).Append(": ").Append(value).Append("\r\n");
        }
        builder.Append("Transfer-Encoding: chunked\r\nConnection: close\r\n\r\n");
        await stream.WriteAsync(Encoding.ASCII.GetBytes(builder.ToString()), cancellationToken).ConfigureAwait(false);
    }

    private static async Task WriteChunkAsync(NetworkStream stream, ReadOnlyMemory<byte> data, CancellationToken cancellationToken)
    {
        await stream.WriteAsync(Encoding.ASCII.GetBytes(data.Length.ToString("X") + "\r\n"), cancellationToken).ConfigureAwait(false);
        await stream.WriteAsync(data, cancellationToken).ConfigureAwait(false);
        await stream.WriteAsync(Encoding.ASCII.GetBytes("\r\n"), cancellationToken).ConfigureAwait(false);
    }

    private static async Task WriteErrorAsync(NetworkStream stream, int status, string message, CancellationToken cancellationToken)
    {
        var safe = message.Replace('"', '\'');
        var body = Encoding.UTF8.GetBytes("{\"error\":\"" + safe + "\"}");
        var header = Encoding.ASCII.GetBytes($"HTTP/1.1 {status} Error\r\nContent-Type: application/json\r\nContent-Length: {body.Length}\r\nConnection: close\r\n\r\n");
        await stream.WriteAsync(header, cancellationToken).ConfigureAwait(false);
        await stream.WriteAsync(body, cancellationToken).ConfigureAwait(false);
    }

    private static readonly HashSet<string> HopByHop = new(StringComparer.OrdinalIgnoreCase)
    {
        "Connection", "Keep-Alive", "Proxy-Authenticate", "Proxy-Authorization", "TE", "Trailer", "Transfer-Encoding", "Upgrade"
    };

    public void Dispose()
    {
        Stop();
        _http.Dispose();
        _stop.Dispose();
    }
}
